File: fibonacci_series.py
# A Fibonacci series in  is written as following:
# 0, 1, 1, 2, 3, 5, 8, 13, 21, 34 and so on.
# As we have seen, the Fibonacci sequence is the integer sequence in which the first two terms are 0 and 1. After that, the following term is defined as the sum of the previous two terms. 
# Having this piece of information, let's start to code the Fibonacci series in .

# note: both fibonacci_series() and fibonacci_generator() do the same thing but fibonacci_generator is space efficient as it outputs a stream and so does not need to save all the numbers in a variable
def fibonacci_series(number_of_terms) -> None:
    n1 = 0
    n2 = 1 
    for i in range(number_of_terms):
        next_term = n1 + n2
        n1 = n2
        n2 = next_term
        print(n1)

number_of_terms = 10        
# fibonacci_series(number_of_terms)

# fibonacci_series implemented using generator function, it saves memomory for large data sets
def fibonacci_generator(n):
    a, b = 0, 1
    count = 0
    while count < n:
        yield a
        a, b = b, a + b
        count += 1

# Generate Fibonacci numbers lazily using a generator
fibonacci_gen = fibonacci_generator(5)

# Use the generated Fibonacci numbers
for num in fibonacci_gen:
    print(num)

File: fibonacciSeries.js
// A Fibonacci series in JavaScript is written as following:
// 0, 1, 1, 2, 3, 5, 8, 13, 21, 34 and so on.

// As we have seen, the Fibonacci sequence is the integer sequence in which the first two terms are 0 and 1. After that, the following term is defined as the sum of the previous two terms. 
// Having this piece of information, let's start to code the Fibonacci series in JavaScript.

// The first step is to create a function that will receive the number of terms that we want to generate.
// The second step is to create a variable to store the first term, which is 0.
// The third step is to create a variable to store the second term, which is 1.
// The fourth step is to create a variable to store the next term, which is the sum of the previous two terms.
// The fifth step is to create a loop that will iterate through the number of terms that we want to generate.
// The sixth step is to print the first and second terms.
// The seventh step is to print the next term, which is the sum of the previous two terms.
// The eighth step is to update the first term with the value of the second term.
// The ninth step is to update the second term with the value of the next term.
// The tenth step is to repeat the process until we reach the number of terms that we want to generate.

let n1 = 0, n2 = 1, nextTerm;
let numberOfTerms = 10;

for(let i =0; i < numberOfTerms; i+=1) {
    nextTerm = n1 + n2;
    n1 = n2;
    n2 = nextTerm;
   
    console.log('fibonacci series:', n1);
}